### DISCUSSION TITLE
asyncio usage of ContextVar with await_only

### DISCUSSION THREAD

**Discussion by al-dpopowich** [sqlalchemy/sqlalchemy#7937] [Usage Questions]
On rare occasion I need to register a callback to be executed after a flush using `after_flush_postexec` event handling. Typically, the knowledge of needing to do this is very conditional, _when creating a new X instance when the sky is blue..._, buried deep in model business logic.

Back in synchronous, multithreaded applications I would use `threading.local` to hold a queue (`flush_data = threading.local()` at the module level) and initialize the queue in my `before_flush` event:

```python
flush_data.after_flush_queue = []
```

Then various calls occurring across my models in `before_flush` and `after_flush` event handling may append tuples, `(fn, args, kwargs)` to the queue (throwing an `AttributeError` if `after_flush_queue` does not exist, to detect registering a callback outside of a flush).  My `after_flush_postexec` event handler would:

```python
for fn, args, kwargs in flush_data.after_flush_queue:
   fn(*args, **kwargs)
del flush_data.after_flush_queue
```

Worked like a charm.  

But now I'm trying to duplicate this in async code.  In place of `threading.local` I'm using `contextvars.ContextVar` as is appropriate, but when my `before_event` handler calls an async method (using `await_only()` as discussed in #7857), the async call gets a *copy* of the current context, so if tries to register a callback I get `LookupError` exceptions.  Is there a way to run `await_only()` and have it use _the current context_?

The following demonstrates the problem (tested with 1.4.35):

```python
import asyncio
from contextvars import ContextVar
from sqlalchemy.util import await_only, greenlet_spawn

callback_queue = ContextVar(f'{__name__}.callback_queue')

def register_callback(awaitable):
   """register callback to be called after flush"""
   try:
      queue = callback_queue.get()
      if queue is None:
         raise ValueError
   except (LookupError, ValueError):
      raise RuntimeError('Programming error: calling register_callback() outside of flush')

   print(f'>>>>>>>>>>> appending to queue: {awaitable}')
   queue.append(awaitable)

async def other_register():
   """Some async function called via await_only that attempts to register a callback"""
   print('>>>>>>>>>>> register_callback(callback(id=2))')
   register_callback(callback(id=2))

async def callback(*, id):
   """Simulated Callback"""
   print('INSIDE CALLBACK', id)

def simulate_before_flush():
   """Simulating flush event"""
   # initialize queue
   callback_queue.set([])

   # This will work
   print('>>>>>>>>>>> register_callback(callback(id=1))')
   register_callback(callback(id=1))

   # but if I call out to an async function that adds to the queue, it
   # blows up with a LookupError exception.  Comment-out the 
   # following to see the mechanics work as expected.
   await_only(other_register())

def simulate_after_flush_postexec():
   """Simulate flush event"""
   queue = callback_queue.get()
   for awaitable in queue:
      await_only(awaitable)

   # "empty" queue
   callback_queue.set(None)


def main():
   """main entry point to simulate flush events"""
   simulate_before_flush()
   simulate_after_flush_postexec()

asyncio.run(greenlet_spawn(main))
```

**Comment by zzzeek** [comment]
Why not use [Session.info](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session%20info#sqlalchemy.orm.Session.info) for this context instead?  The `Session` is what stays the same throughout the course of a session operation (AsyncSession included) and the purpose of this collection is to be able to track things between usage and events.

**Reply by zzzeek** [reply]
as far as the ContextVar situation I have no idea.   I thought we had ContextVar doing all the right things, so not sure if the issue you see is local to the usage of ContextVar inside of a run_sync(), which is not part of our tested coverage.  I'm not familiar with what happens in that case or if it could be changed.

**Reply by al-dpopowich** [reply]
> Why not use [Session.info](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session%20info#sqlalchemy.orm.Session.info) for this context instead? 

That could work _if_ I have a handle to the current `session`.  Referring to usage suggested in #7857, I could use `session.info` because I do pass the session:

```python
class MyModel(Base):
   ...
   async def on_post_create(self, session):
      if await self.some_condition():
         register_callback(session, self.some_awaitable(42))
```

But if knowledge of needing to take some action is deep inside business logic, I'll need to propagate that knowledge somehow.  I can probably get by with `session.info` for now.  Meanwhile...

> as far as the ContextVar situation I have no idea. I thought we had ContextVar doing all the right things.

I'm not familiar with `greenlet` nor `sqlalchemy`'s usage of it enough to know how it's using `ContextVar`s, but this does seem broken to me.  I clearly set the var, then make async calls to other functions which see an unset var, generating the LookupError.  If I can spare some cycles I'll dig into this more.

**Reply by CaselIT** [reply]
Interesting it seems that greenlet pre 0.4.17 and our post behaviour is different. The pre version seems to work correctly in this case.

Note that the current behavior is the expected one, meaning that it does what it was designed to do, meaning that the functions run in a copy of the context at the point of execution of greenlet_spawn.
What's not supported is setting the context var variable from inside the functions.

See the current test for an example of the current behavior: 
https://github.com/sqlalchemy/sqlalchemy/blob/dd414e7eaba595a333e593865196de1fd91fe6db/test/base/test_concurrency_py3k.py#L143-L170

I'm not immediately sure how or if this is fixable

**Reply by zzzeek** [reply]
does that test fail if you use the pre 0.4.17 greenlet?

**Reply by CaselIT** [reply]
no, that test work correctly

**Reply by CaselIT** [reply]
Basically what does not work with greenlet > 0.4.17 but works with greenlet < 0.4.17 is this test:
```py
    @async_test
    async def test_contextvars(self):
        import asyncio
        import contextvars

        var = contextvars.ContextVar("var")
        concurrency = 5

        async def async_inner(val):
            eq_(val, var.get())
            return var.get()

        def inner(val):
            retval = await_only(async_inner(val))
            eq_(val, var.get())
            eq_(retval, val)

            var.set(val + 1)  # this set is currently not supported
            retval = await_only(async_inner(val + 1))
            eq_(val + 1, var.get())
            eq_(retval, val + 1)

            return retval

        async def task(val):
            var.set(val)
            return await greenlet_spawn(inner, val)

        values = {
            await coro
            for coro in asyncio.as_completed(
                [task(i) for i in range(concurrency)]
            )
        }
        # eq_(values, set(range(concurrency)))
        eq_(values, set(range(1, concurrency + 1)))
```

**Reply by zzzeek** [reply]
so....this is not the same thing they are doing above then right

**Reply by CaselIT** [reply]
I can't seems to find anything wrong with this change
```diff
diff --git a/lib/sqlalchemy/util/_concurrency_py3k.py b/lib/sqlalchemy/util/_concurrency_py3k.py
index 6ad099eef..477db8f9e 100644
--- a/lib/sqlalchemy/util/_concurrency_py3k.py
+++ b/lib/sqlalchemy/util/_concurrency_py3k.py
@@ -76,7 +76,7 @@ class _AsyncIoGreenlet(greenlet):  # type: ignore
         greenlet.__init__(self, fn, driver)
         self.driver = driver
         if _copy_context is not None:
-            self.gr_context = _copy_context()
+            self.gr_context = driver.gr_context


 def await_only(awaitable: Awaitable[_T]) -> _T:
```

This fixes the test above, and the OP example. @fantix do you have suggestions here?

**Reply by CaselIT** [reply]
> so....this is not the same thing they are doing above then right

no, the change to test_contextvars does the same as the opening example, namely sets the context var inside the greenlet context.

I think the issue here is that _copy_context returns a copy of the greenlet_spawn context, so sync code runs in the copied context, while async context runs in the original greenlet_spawn context

**Reply by CaselIT** [reply]
The bot does not seem to comment on discussion. Proposed a change here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3791
